Client/Frontend/Sync/BraveCore/BraveSyncQRCodeModel.swift

```diff
@@ -0,0 +1,142 @@
+// Copyright 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import BraveShared
+import Shared
+import BraveCore
+
+private let log = Logger.braveCoreLogger
+
+enum BraveSyncQRCodeError {
+    case none
+    case insecure
+    case newerVersion
+    case olderVersion
+    case invalidFormat
+    case expired
+    case futureDate
+}
+
+struct BraveSyncQRCodeModel: Codable {
+    let version: Int
+    let syncHexCode: String
+    let notValidAfter: Int64
+    
+    /// The currently supported version of QR code
+    static let currentlySupportedVersion = 2
+    
+    /// The duration for which a QR code is valid
+    static let validityDuration = 60.minutes
+    
+    /// TimeInterval of BraveSync Epoch
+    /// 2021-11-02T00:00:00+0000 [yyyy-MM-dd'T'hh:mm:ssZ]
+    static let braveSyncEpoch = 1635811200
+    
+    /// TimeInterval of Brave Sync Sunset
+    /// 2022-01-01T00:00:00+0000 [yyyy-MM-dd'T'hh:mm:ssZ]
+    static let sunsetDate = 1640995200
+    
+    init(syncHexCode: String) {
+        self.version = BraveSyncQRCodeModel.currentlySupportedVersion
+        self.syncHexCode = syncHexCode
+        self.notValidAfter = Int64(Date().addingTimeInterval(BraveSyncQRCodeModel.validityDuration).timeIntervalSince1970)
+    }
+    
+    init(version: Int, syncHexCode: String, notValidAfter: TimeInterval) {
+        self.version = version
+        self.syncHexCode = syncHexCode
+        self.notValidAfter = Int64(notValidAfter)
+    }
+    
+    private enum CodingKeys: String, CodingKey {
+        case version
+        case syncHexCode = "sync_code_hex"
+        case notValidAfter = "not_after"
+    }
+    
+    var jsonData: Data? {
+        do {
+            return try JSONEncoder().encode(self)
+        } catch {
+            log.error("Error converting QR Code to JSON: \(error)")
+        }
+        return nil
+    }
+    
+    var jsonString: String? {
+        if let data = jsonData {
+            return String(data: data, encoding: .utf8)
+        }
+        return nil
+    }
+    
+    static func from(string: String) -> BraveSyncQRCodeModel? {
+        if let data = string.data(using: .utf8) {
+            do {
+                return try JSONDecoder().decode(Self.self, from: data)
+            } catch {
+                log.error("Error converting QR Code to BraveSyncQRCodeModel: \(error)")
+            }
+        }
+        return nil
+    }
+    
+    func validate(syncAPI: BraveSyncAPI) -> BraveSyncQRCodeError {
+        // Older verison of the QR code (not backwards compatible)
+        if version < BraveSyncQRCodeModel.currentlySupportedVersion {
```

Comment:
Is this for future proofing, why do we check the version as smaller or bigger. Cant we use it as .notSupported? 
Do we have different error messages for these cases here ?

Reply:
It is. We don't want to support any QR-Codes less than the current version "2". So if the version is "1", we want an error (For example, QR code version is insecure or super old).

If the version is > the current supported version, we want a different error (For example, user might need to update their app to scan a newer QR code version).